src/message_log.rs

```diff
@@ -131,131 +118,89 @@ impl PbftLog {
+        // Check if Prepared predicate is true
+        if !self.check_prepared(info, f)? {
+            return Ok(false);
         }
-        self.check_msg_against_log(&pbft_message, true, 2 * f + 1)?;
 
-        self.check_prepared(&pbft_message.as_msg_type(PbftMessageType::Prepare), f)?;
-        Ok(())
+        // Check if we have 2f + 1 matching Commits
+        let preprep_msg = self
+            .get_one_msg(info, &PbftMessageType::PrePrepare)
+            .unwrap();
+        Ok(self.log_has_required_msgs(&PbftMessageType::Commit, &preprep_msg, true, 2 * f + 1))
     }
 
-    /// Check an incoming message against its counterparts in the message log
-    pub fn check_msg_against_log(
+    /// Get one message matching the type, view number, and sequence number
+    fn get_one_msg(
         &self,
-        message: &ParsedMessage,
-        check_match: bool,
-        num_cutoff: u64,
-    ) -> Result<(), PbftError> {
-        let info = message.info();
-        let msg_type = PbftMessageType::from(info.get_msg_type());
-
-        let msg_infos: Vec<&PbftMessageInfo> =
-            self.get_message_infos(&msg_type, info.get_seq_num(), info.get_view());
-
-        let num_cp_msgs = num_unique_signers(&msg_infos);
-        if num_cp_msgs < num_cutoff {
-            return Err(PbftError::WrongNumMessages(
-                msg_type,
-                num_cutoff as usize,
-                num_cp_msgs as usize,
-            ));
+        info: &PbftMessageInfo,
+        msg_type: &PbftMessageType,
+    ) -> Option<&ParsedMessage> {
+        let msgs = self.get_messages_of_type(msg_type, info.get_seq_num(), info.get_view());
+        if msgs.is_empty() {
+            None
+        } else {
+            Some(msgs[0])
         }
+    }
 
-        if check_match {
-            let non_matches: usize = msg_infos
-                .iter()
-                .filter(|&m| !infos_match(message.info(), m))
-                .count();
-            if non_matches > 0 {
-                return Err(PbftError::MessageMismatch(msg_type));
-            }
-        }
-        Ok(())
+    /// Check if the log contains `required` number of messages with type `msg_type` that match the
+    /// sequence and view number of the provided `ref_msg`, as well as its block (optional)
+    pub fn log_has_required_msgs(
+        &self,
+        msg_type: &PbftMessageType,
+        ref_msg: &ParsedMessage,
+        check_block: bool,
+        required: u64,
+    ) -> bool {
+        let msgs = self.get_messages_of_type(
+            msg_type,
+            ref_msg.info().get_seq_num(),
+            ref_msg.info().get_view(),
+        );
+
+        let msgs = if check_block {
+            msgs.iter()
+                .filter(|msg| msg.get_block() == ref_msg.get_block())
+                .cloned()
+                .collect()
+        } else {
+            msgs
+        };
+
+        num_unique_signers(msgs) >= required
```

Comment:
I agree with this and have been eliminating these kinds of functions as I've been improving things. I actually just removed the function altogether. The log doesn't allow duplicates, and at this point we only have blocks that match the sequence, view, type, and block; that means the only variable between these messages will be the signer id, and since there are no duplicates, all of these messages are from unique signers.